server/src/com/cloud/storage/VolumeApiServiceImpl.java

```diff
@@ -1259,92 +1261,139 @@ private VolumeVO orchestrateResizeVolume(long volumeId, long currentSize, long n
         }
     }
 
-    @Override
     @DB
+    @Override
     @ActionEvent(eventType = EventTypes.EVENT_VOLUME_DELETE, eventDescription = "deleting volume")
-    public boolean deleteVolume(long volumeId, Account caller) throws ConcurrentOperationException {
+    /**
+     * Executes the removal of the volume. If the volume is only allocated we do not try to remove it from primary and secondary storage.
+     * Otherwise, after the removal in the database, we will try to remove the volume from both primary and secondary storage.
+     */
+    public boolean deleteVolume(long volumeId, Account caller){
+        VolumeVO volume = retrieveAndValidateVolume(volumeId, caller);
+        try {
+            destroyVolumeIfPossible(volume);
+            // Mark volume as removed if volume has not been created on primary or secondary
+            if (volume.getState() == Volume.State.Allocated) {
+                _volsDao.remove(volumeId);
+                stateTransitTo(volume, Volume.Event.DestroyRequested);
+                return true;
+            }
+            expungeVolumesInPrimaryStorageIfNeeded(volume);
+            expungeVolumesInSecondaryStorageIfNeeded(volume);
+            cleanVolumesCache(volume);
+            return true;
+        } catch (InterruptedException | ExecutionException | NoTransitionException e) {
+            s_logger.warn("Failed to expunge volume: " + volume.getUuid(), e);
+            return false;
+        }
+    }
+
+    /**
```

Comment:
java doc on protected method. why not private and should the javadoc be plain comment?

Reply:
I am using `protected` modifier here because I have unit tests for this method.

I used Java doc notation style because that is what I am used to do. I am change though. It is probably an influence of Apache commons and Spring suite code style.

Reply:
so why not package default? should be good enough for tests

Reply:
That is a good question. Right now, we are not overriding this method in a subclass. Therefore, it can use the default access modifier. Would you like me to change that?

Reply:
ok, I gave you my approval as is, so ...
for my sake the methods should be private, they were part of the bigger method so could not be overridden. also the comment should be part of the caller. the callee already has the code to examine. €0,02